File: src/apps/edge_computing.cpp
#include <simgrid/s4u.hpp>
#include <iostream>
#include <string>

XBT_LOG_NEW_DEFAULT_CATEGORY(edge_computing, "Edge Computing Application");

namespace sg4 = simgrid::s4u;

/**
 * @brief Edge processing application
 * 
 * Simulates Edge devices that process data locally and send
 * results to a gateway when necessary.
 */

// Actor for Edge device
class EdgeDevice {
    std::string gateway_name;
    double computation_size;
    double data_size;
    
public:
    EdgeDevice(const std::string& gateway, double comp_size, double data_size_)
        : gateway_name(gateway), computation_size(comp_size), data_size(data_size_) {}
    
    void operator()() {
        sg4::Host* this_host = sg4::this_actor::get_host();
        XBT_INFO("Edge Device '%s' started (speed: %.2f Gf)", 
                 this_host->get_cname(), this_host->get_speed() / 1e9);
        
        // Phase 1: Data collection
        XBT_INFO("Collecting sensor data...");
        sg4::this_actor::sleep_for(1.0);
        
        // Phase 2: Local processing
        XBT_INFO("Processing data locally (%.2f MFlops)...", computation_size / 1e6);
        sg4::this_actor::execute(computation_size);
        
        // Phase 3: Send results to gateway
        XBT_INFO("Sending results to gateway (%s)...", gateway_name.c_str());
        
        auto* mbox = sg4::Mailbox::by_name(gateway_name);
        auto* payload = new std::string("Processed data from " + 
                                        std::string(this_host->get_cname()));
        mbox->put(payload, data_size);
        
        XBT_INFO("Task completed");
    }
};

// Actor for Edge Gateway
class EdgeGateway {
    int num_devices;
    
public:
    EdgeGateway(int num) : num_devices(num) {}
    
    void operator()() {
        sg4::Host* this_host = sg4::this_actor::get_host();
        XBT_INFO("Edge Gateway '%s' started", this_host->get_cname());
        
        auto* mbox = sg4::Mailbox::by_name(this_host->get_cname());
        
        // Receive data from all devices
        for (int i = 0; i < num_devices; i++) {
            auto* msg = mbox->get<std::string>();
            XBT_INFO("Received: %s", msg->c_str());
            delete msg;
            
            // Process aggregated data
            sg4::this_actor::execute(5e8);  // 500 MFlops
        }
        
        XBT_INFO("All data processed. Gateway completed.");
    }
};

int main(int argc, char* argv[]) {
    sg4::Engine e(&argc, argv);
    
    // Verify arguments
    if (argc < 2) {
        XBT_CRITICAL("Usage: %s <platform_file.xml>", argv[0]);
        XBT_CRITICAL("Example: %s platforms/edge_platform.xml", argv[0]);
        return 1;
    }
    
    // Load platform
    e.load_platform(argv[1]);
    
    // Get hosts
    std::vector<sg4::Host*> hosts = e.get_all_hosts();
    
    if (hosts.empty()) {
        XBT_CRITICAL("No hosts found in platform");
        return 1;
    }
    
    XBT_INFO("=== Edge Computing Application ===");
    XBT_INFO("Platform loaded with %zu hosts", hosts.size());
    
    // Find gateway (assume it's the first or has "gateway" in the name)
    sg4::Host* gateway = nullptr;
    std::vector<sg4::Host*> devices;
    
    for (auto* host : hosts) {
        std::string name = host->get_cname();
        if (name.find("gateway") != std::string::npos) {
            gateway = host;
        } else {
            devices.push_back(host);
        }
    }
    
    // If no explicit gateway, use first host
    if (!gateway && !hosts.empty()) {
        gateway = hosts[0];
        devices.clear();
        for (size_t i = 1; i < hosts.size(); i++) {
            devices.push_back(hosts[i]);
        }
    }
    
    if (!gateway || devices.empty()) {
        XBT_CRITICAL("At least one gateway and one device are required");
        return 1;
    }
    
    XBT_INFO("Gateway: %s", gateway->get_cname());
    XBT_INFO("Edge Devices: %zu", devices.size());
    
    // Create gateway actor
    gateway->add_actor("gateway", EdgeGateway(devices.size()));
    
    // Create actors for edge devices
    for (size_t i = 0; i < devices.size(); i++) {
        double comp_size = 1e9;  // 1 GFlop
        double data_size = 1e6;  // 1 MB
        
        devices[i]->add_actor("device_" + std::to_string(i),
                              EdgeDevice(gateway->get_cname(), comp_size, data_size));
    }
    
    // Run simulation
    e.run();
    
    XBT_INFO("=== Simulation completed ===");
    XBT_INFO("Simulated time: %.2f seconds", sg4::Engine::get_clock());
    
    return 0;
}

File: src/python/tests/edge_computing.py
"""
edge_computing.py – Simulates edge devices that process data locally and
forward results to a central gateway.

Mirrors tests/edge_computing.cpp.

Usage
-----
    python edge_computing.py <platform_file.xml>
"""
from __future__ import annotations

import sys
import os

_HERE = os.path.dirname(os.path.abspath(__file__))
_ROOT = os.path.join(_HERE, "..")
if _ROOT not in sys.path:
    sys.path.insert(0, _ROOT)

import simgrid


class EdgeDevice:
    """Senses data, processes locally, sends results to a gateway mailbox."""

    def __init__(
        self,
        gateway_name: str,
        computation_size: float = 2e8,
        data_size: float = 1e5,
    ) -> None:
        self._gateway      = gateway_name
        self._computation  = computation_size   # Flops
        self._data_size    = data_size           # bytes

    def __call__(self) -> None:
        host = simgrid.this_actor.get_host()
        simgrid.this_actor.info(
            f"Edge Device '{host.name}' started "
            f"(speed: {host.speed / 1e9:.2f} Gf)"
        )

        # Phase 1: data collection
        simgrid.this_actor.info("Collecting sensor data…")
        simgrid.this_actor.sleep_for(1.0)

        # Phase 2: local processing
        simgrid.this_actor.info(
            f"Processing data locally ({self._computation / 1e6:.2f} MFlops)…"
        )
        simgrid.this_actor.execute(self._computation)

        # Phase 3: send results to gateway
        simgrid.this_actor.info(f"Sending results to gateway ({self._gateway})…")
        mbox = simgrid.Mailbox.by_name(self._gateway)
        mbox.put(f"Processed data from {host.name}", int(self._data_size))

        simgrid.this_actor.info("Task completed")


class EdgeGateway:
    """Collects results from all edge devices and performs aggregation."""

    def __init__(self, num_devices: int) -> None:
        self._num_devices = num_devices

    def __call__(self) -> None:
        host = simgrid.this_actor.get_host()
        simgrid.this_actor.info(f"Edge Gateway '{host.name}' started")

        mbox = simgrid.Mailbox.by_name(host.name)
        for i in range(self._num_devices):
            msg = mbox.get()
            simgrid.this_actor.info(f"Received: {msg}")
            simgrid.this_actor.execute(5e8)   # 500 MFlops per aggregation

        simgrid.this_actor.info("All data processed. Gateway completed.")


def main() -> None:
    e = simgrid.Engine(sys.argv)

    if len(sys.argv) < 2:
        print(f"Usage: {sys.argv[0]} <platform_file.xml>", file=sys.stderr)
        sys.exit(1)

    e.load_platform(sys.argv[1])
    hosts = e.all_hosts

    if not hosts:
        print("No hosts found in platform", file=sys.stderr)
        sys.exit(1)

    simgrid.this_actor.info("=== Edge Computing Application ===")
    simgrid.this_actor.info(f"Platform loaded with {len(hosts)} hosts")

    # Identify gateway by name; fall back to last host if not found
    gateway_host = next((h for h in hosts if "gateway" in h.name), hosts[-1])
    device_hosts = [h for h in hosts if h is not gateway_host]

    if not device_hosts:
        print("At least 2 hosts required (1 gateway + 1 device).", file=sys.stderr)
        sys.exit(1)

    # Deploy gateway
    simgrid.Actor.create("edge_gateway", gateway_host,
                         EdgeGateway(len(device_hosts)))

    # Deploy edge devices
    for host in device_hosts:
        simgrid.Actor.create("edge_device", host,
                             EdgeDevice(gateway_host.name, 2e8, 1e5))

    e.run()
    simgrid.this_actor.info("=== Simulation completed ===")
    simgrid.this_actor.info(f"Simulated time: {simgrid.Engine.clock:.2f} s")


if __name__ == "__main__":
    main()
